File: packages/clients/python/tangram/blob.py
"""Blob class for Tangram.

This module mirrors the JavaScript packages/clients/js/src/blob.ts.
"""

from __future__ import annotations

import base64
from typing import Any, Dict, List, Optional, Union

from tangram.object_ import Object_


async def blob(*args: Any) -> "Blob":
    """Create a new blob."""
    return await Blob.new(*args)


class Blob:
    """A blob is a binary large object."""

    _state: Object_.State

    def __init__(
        self,
        *,
        id: Optional[str] = None,
        object: Optional[Dict[str, Any]] = None,
        stored: bool = False,
    ) -> None:
        """Initialize a blob."""
        obj = {"kind": "blob", "value": object} if object is not None else None
        self._state = Object_.State(id=id, object=obj, stored=stored)

    @property
    def state(self) -> Object_.State:
        """Get the object state."""
        return self._state

    @classmethod
    def with_id(cls, id: str) -> "Blob":
        """Create a blob with the given ID."""
        return cls(id=id, stored=True)

    @classmethod
    def with_object(cls, object: Dict[str, Any]) -> "Blob":
        """Create a blob with the given object."""
        return cls(object=object, stored=False)

    @classmethod
    def from_data(cls, data: Dict[str, Any]) -> "Blob":
        """Create a blob from serialized data."""
        return cls.with_object(Blob.Object.from_data(data))

    @classmethod
    async def new(cls, *args: Any) -> "Blob":
        """Create a new blob from arguments."""
        import tangram as tg

        arg = await cls.arg(*args)
        children = arg.get("children", [])
        if not children:
            return cls.with_object({"bytes": b""})
        elif len(children) == 1:
            return children[0]["blob"]
        else:
            return cls.with_object({"children": children})

    @classmethod
    async def leaf(cls, *args: Any) -> "Blob":
        """Create a leaf blob from raw data."""
        import tangram as tg

        resolved = await tg.resolve(list(args))
        objects = []
        for arg in resolved:
            if arg is None:
                objects.append(b"")
            elif isinstance(arg, str):
                objects.append(arg.encode("utf-8"))
            elif isinstance(arg, bytes):
                objects.append(arg)
            elif isinstance(arg, Blob):
                objects.append(await arg.bytes())
            else:
                objects.append(b"")
        data = b"".join(objects)
        return cls.with_object({"bytes": data})

    @classmethod
    async def branch(cls, *args: Any) -> "Blob":
        """Create a branch blob."""
        arg = await cls.arg(*args)
        return cls.with_object({"children": arg.get("children", [])})

    @classmethod
    async def arg(cls, *args: Any) -> Dict[str, Any]:
        """Process blob arguments using Args.apply."""
        import tangram as tg

        async def map_fn(arg: Any) -> Dict[str, Any]:
            if arg is None:
                return {"children": []}
            elif isinstance(arg, str):
                data = arg.encode("utf-8")
                blob = Blob.with_object({"bytes": data})
                return {"children": [{"blob": blob, "length": len(data)}]}
            elif isinstance(arg, bytes):
                blob = Blob.with_object({"bytes": arg})
                return {"children": [{"blob": blob, "length": len(arg)}]}
            elif isinstance(arg, Blob):
                length = await arg.length()
                return {"children": [{"blob": arg, "length": length}]}
            elif isinstance(arg, dict):
                return arg
            else:
                return {"children": []}

        return await tg.Args.apply(
            args=list(args),
            map_fn=map_fn,
            reduce_spec={"children": "append"},
        )

    @staticmethod
    def expect(value: Any) -> "Blob":
        """Assert that value is a Blob and return it."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, Blob))
        return value

    @staticmethod
    def assert_(value: Any) -> None:
        """Assert that value is a Blob."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, Blob))

    @property
    def id(self) -> str:
        """Get the blob ID."""
        from tangram.assert_ import assert_

        id = self._state.id
        assert_(Object_.Id.kind(id) == "blob")
        return id

    async def object(self) -> Dict[str, Any]:
        """Get the blob object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "blob")
        return obj.get("value", obj)

    async def load(self) -> Dict[str, Any]:
        """Load the blob object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "blob")
        return obj.get("value", obj)

    def unload(self) -> None:
        """Unload the blob data."""
        self._state.unload()

    async def store(self) -> str:
        """Store the blob and return its ID."""
        import tangram as tg

        await tg.Value.store(self)
        return self.id

    async def children(self) -> List[Object_]:
        """Get child objects."""
        return await self._state.children()

    async def length(self) -> int:
        """Get the blob length."""
        obj = await self.object()
        if "children" in obj:
            return sum(child["length"] for child in obj["children"])
        else:
            return len(obj.get("bytes", b""))

    async def read(
        self,
        position: Optional[int] = None,
        length: Optional[int] = None,
        size: Optional[int] = None,
    ) -> bytes:
        """Read blob contents."""
        import tangram as tg

        id = await self.store()
        arg: Dict[str, Any] = {"blob": id}
        if position is not None:
            arg["position"] = position
        if length is not None:
            arg["length"] = length
        if size is not None:
            arg["size"] = size
        return await tg.handle.read(arg)

    async def bytes(self) -> bytes:
        """Get the blob contents as bytes."""
        return await self.read()

    async def text(self) -> str:
        """Get the blob contents as text."""
        data = await self.bytes()
        return data.decode("utf-8")

    # Nested Object namespace.
    class Object:
        """Namespace for Blob.Object operations."""

        @staticmethod
        def to_data(obj: Dict[str, Any]) -> Dict[str, Any]:
            """Convert object to data."""
            if "bytes" in obj:
                return {"bytes": base64.b64encode(obj["bytes"]).decode("ascii")}
            else:
                return {
                    "children": [
                        {"blob": child["blob"].id, "length": child["length"]}
                        for child in obj.get("children", [])
                    ]
                }

        @staticmethod
        def from_data(data: Dict[str, Any]) -> Dict[str, Any]:
            """Convert data to object."""
            if "bytes" in data:
                return {"bytes": base64.b64decode(data["bytes"])}
            else:
                return {
                    "children": [
                        {"blob": Blob.with_id(child["blob"]), "length": child["length"]}
                        for child in data.get("children", [])
                    ]
                }

        @staticmethod
        def children(obj: Optional[Dict[str, Any]]) -> List[Object_]:
            """Get child objects from blob object."""
            if obj is None:
                return []
            if "children" in obj:
                return [child["blob"] for child in obj["children"]]
            return []


# Type aliases.
Blob.Id = str
Blob.Arg = Union[None, str, bytes, "Blob", Dict[str, Any]]
Blob.Data = Dict[str, Any]

File: packages/clients/js/src/blob.ts
import * as tg from "./index.ts";
import { unindent } from "./template.ts";

/** Create a blob. */
export function blob(
	strings: TemplateStringsArray,
	...placeholders: tg.Args<string>
): tg.Blob.Builder;
export function blob(...args: tg.Args<tg.Blob.Arg>): tg.Blob.Builder;
export function blob(
	firstArg:
		| TemplateStringsArray
		| tg.Unresolved<tg.ValueOrMaybeMutationMap<tg.Blob.Arg>>,
	...args: tg.Args<tg.Blob.Arg>
): tg.Blob.Builder {
	return new tg.Blob.Builder(firstArg, ...args);
}

export class Blob {
	#state: tg.Object.State;

	constructor(arg: {
		id?: tg.Blob.Id;
		object?: tg.Blob.Object;
		stored: boolean;
	}) {
		let object =
			arg.object !== undefined
				? { kind: "blob" as const, value: arg.object }
				: undefined;
		this.#state = new tg.Object.State({
			id: arg.id,
			object,
			stored: arg.stored,
		});
	}

	get state(): tg.Object.State {
		return this.#state;
	}

	/** Get a blob with an ID. */
	static withId(id: tg.Blob.Id): tg.Blob {
		return new tg.Blob({ id, stored: true });
	}

	static withObject(object: tg.Blob.Object): tg.Blob {
		return new tg.Blob({ object, stored: false });
	}

	static fromData(data: tg.Blob.Data): tg.Blob {
		return tg.Blob.withObject(tg.Blob.Object.fromData(data));
	}

	/** Create a blob. */
	static async new(...args: tg.Args<tg.Blob.Arg>): Promise<tg.Blob> {
		let arg = await tg.Blob.arg(...args);
		let blob: tg.Blob;
		if (!arg.children || arg.children.length === 0) {
			blob = tg.Blob.withObject({ bytes: new Uint8Array() });
		} else if (arg.children.length === 1) {
			blob = arg.children[0]!.blob;
		} else {
			blob = tg.Blob.withObject({ children: arg.children });
		}
		return blob;
	}

	static leaf(
		...args: tg.Args<undefined | string | Uint8Array | tg.Blob>
	): tg.Blob.Builder {
		return tg.Blob.Builder.leaf(...args);
	}

	static branch(...args: tg.Args<tg.Blob.Arg>): tg.Blob.Builder {
		return tg.Blob.Builder.branch(...args);
	}

	static async arg(...args: tg.Args<tg.Blob.Arg>): Promise<tg.Blob.Arg.Object> {
		return await tg.Args.apply({
			args,
			map: async (arg) => {
				if (arg === undefined) {
					return { children: [] };
				} else if (typeof arg === "string") {
					let bytes = tg.encoding.utf8.encode(arg);
					let blob = tg.Blob.withObject({ bytes });
					let length = bytes.length;
					return { children: [{ blob, length }] };
				} else if (arg instanceof Uint8Array) {
					let bytes = arg;
					let blob = tg.Blob.withObject({ bytes });
					let length = bytes.length;
					return { children: [{ blob, length }] };
				} else if (arg instanceof tg.Blob) {
					let length = await arg.length;
					let child = { blob: arg, length };
					return {
						children: [child],
					};
				} else {
					return arg;
				}
			},
			reduce: {
				children: "append",
			},
		});
	}

	/** Expect that a value is a `tg.Blob`. */
	static expect(value: unknown): tg.Blob {
		tg.assert(value instanceof tg.Blob);
		return value;
	}

	/** Assert that a value is a `tg.Blob`. */
	static assert(value: unknown): asserts value is tg.Blob {
		tg.assert(value instanceof tg.Blob);
	}

	/** Get this blob's ID. */
	get id(): tg.Blob.Id {
		let id = this.#state.id;
		tg.assert(tg.Object.Id.kind(id) === "blob");
		return id;
	}

	async object(): Promise<tg.Blob.Object> {
		let object = await this.#state.load();
		tg.assert(object.kind === "blob");
		return object.value;
	}

	async load(): Promise<tg.Blob.Object> {
		let object = await this.#state.load();
		tg.assert(object.kind === "blob");
		return object.value;
	}

	unload(): void {
		this.#state.unload();
	}

	/** Store this blob. */
	async store(): Promise<tg.Blob.Id> {
		await tg.Value.store(this);
		return this.id;
	}

	get children(): Promise<Array<tg.Object>> {
		return this.#state.children;
	}

	/** Get this blob's length. */
	get length(): Promise<number> {
		return (async () => {
			let object = await this.object();
			if ("children" in object) {
				return object.children
					.map(({ length }) => length)
					.reduce((a, b) => a + b, 0);
			} else {
				return object.bytes.byteLength;
			}
		})();
	}

	/** Read from this blob. */
	async read(options?: tg.Blob.ReadOptions): Promise<Uint8Array> {
		let id = await this.store();
		const arg = { blob: id, ...options };
		return await tg.handle.read(arg);
	}

	/** Read this entire blob to a `Uint8Array`. */
	get bytes(): Promise<Uint8Array> {
		return this.read();
	}

	/** Read this entire blob to a string. */
	get text(): Promise<string> {
		return (async () => {
			return tg.encoding.utf8.decode(await this.bytes);
		})();
	}
}

export namespace Blob {
	export type Id = string;

	export class Builder {
		#args: Array<any>;
		#create: (...args: Array<any>) => Promise<tg.Blob>;

		constructor(
			raw: boolean,
			strings: TemplateStringsArray,
			...placeholders: tg.Args<string>
		);
		constructor(
			strings: TemplateStringsArray,
			...placeholders: tg.Args<string>
		);
		constructor(
			firstArg:
				| TemplateStringsArray
				| tg.Unresolved<tg.ValueOrMaybeMutationMap<tg.Blob.Arg>>,
			...args: tg.Args<tg.Blob.Arg>
		);
		constructor(...args: tg.Args<tg.Blob.Arg>);
		constructor(...args: any[]) {
			let raw = false;
			if (typeof args[0] === "boolean") {
				raw = args[0];
				args = args.slice(1);
			}
			let firstArg = args[0];
			if (Array.isArray(firstArg) && "raw" in firstArg) {
				let strings = firstArg as TemplateStringsArray;
				let placeholders = args.slice(1) as tg.Args<string>;
				let components = [];
				for (let i = 0; i < strings.length - 1; i++) {
					let string = strings[i]!;
					components.push(string);
					let placeholder = placeholders[i]!;
					components.push(placeholder);
				}
				components.push(strings[strings.length - 1]!);
				let string = components.join("");
				if (!raw) {
					string = unindent([string]).join("");
				}
				this.#args = [string];
			} else {
				this.#args = args;
			}
			this.#create = tg.Blob.new;
		}

		static leaf(
			...args: tg.Args<undefined | string | Uint8Array | tg.Blob>
		): tg.Blob.Builder {
			let builder = new tg.Blob.Builder();
			builder.#args = args;
			builder.#create = tg.Blob.Builder.createLeaf;
			return builder;
		}

		static async createLeaf(
			...args: tg.Args<undefined | string | Uint8Array | tg.Blob>
		): Promise<tg.Blob> {
			let resolved = await Promise.all(args.map(tg.resolve));
			let objects = await Promise.all(
				resolved.map(async (arg) => {
					if (arg === undefined) {
						return new Uint8Array();
					} else if (typeof arg === "string") {
						return tg.encoding.utf8.encode(arg);
					} else if (arg instanceof Uint8Array) {
						return arg;
					} else {
						return await arg.bytes;
					}
				}),
			);
			let length = objects.reduce(
				(length, bytes) => length + bytes.byteLength,
				0,
			);
			let bytes = new Uint8Array(length);
			let offset = 0;
			for (let entry of objects) {
				bytes.set(entry, offset);
				offset += entry.byteLength;
			}
			let object = { bytes };
			return tg.Blob.withObject(object);
		}

		static branch(...args: tg.Args<tg.Blob.Arg>): tg.Blob.Builder {
			let builder = new tg.Blob.Builder();
			builder.#args = args;
			builder.#create = tg.Blob.Builder.createBranch;
			return builder;
		}

		static async createBranch(...args: tg.Args<tg.Blob.Arg>): Promise<tg.Blob> {
			let arg = await tg.Blob.arg(...args);
			return tg.Blob.withObject({ children: arg.children ?? [] });
		}

		then<TResult1 = tg.Blob, TResult2 = never>(
			onfulfilled?:
				| ((value: tg.Blob) => TResult1 | PromiseLike<TResult1>)
				| undefined
				| null,
			onrejected?:
				| ((reason: any) => TResult2 | PromiseLike<TResult2>)
				| undefined
				| null,
		): PromiseLike<TResult1 | TResult2> {
			return this.#create(...this.#args).then(onfulfilled, onrejected);
		}
	}

	export type Arg =
		| undefined
		| string
		| Uint8Array
		| tg.Blob
		| tg.Blob.Arg.Object;

	export namespace Arg {
		export type Object = {
			children?: Array<tg.Blob.Child> | undefined;
		};
	}

	export type Object = tg.Blob.Leaf | tg.Blob.Branch;

	export namespace Object {
		export let toData = (object: tg.Blob.Object): tg.Blob.Data => {
			if ("bytes" in object) {
				return {
					bytes: tg.encoding.base64.encode(object.bytes),
				};
			} else {
				return {
					children: object.children.map((child) => ({
						blob: child.blob.id,
						length: child.length,
					})),
				};
			}
		};

		export let fromData = (data: tg.Blob.Data): tg.Blob.Object => {
			if ("bytes" in data) {
				return {
					bytes: tg.encoding.base64.decode(data.bytes),
				};
			} else {
				return {
					children: data.children.map((child) => ({
						blob: tg.Blob.withId(child.blob),
						length: child.length,
					})),
				};
			}
		};

		export let children = (object: tg.Blob.Object): Array<tg.Object> => {
			if ("children" in object) {
				return object.children.map(({ blob }) => blob);
			} else {
				return [];
			}
		};
	}

	export type Leaf = {
		bytes: Uint8Array;
	};

	export type Branch = {
		children: Array<tg.Blob.Child>;
	};

	export type Child = {
		blob: tg.Blob;
		length: number;
	};

	export type Data = tg.Blob.Data.Leaf | tg.Blob.Data.Branch;

	export namespace Data {
		export type Leaf = {
			bytes: string;
		};

		export type Branch = {
			children: Array<tg.Blob.Data.Child>;
		};

		export type Child = {
			blob: tg.Blob.Id;
			length: number;
		};

		export let children = (data: tg.Blob.Data): Array<tg.Object.Id> => {
			if ("children" in data) {
				return data.children.map((child) => child.blob);
			} else {
				return [];
			}
		};
	}

	export type ReadOptions = {
		position?: number | string | undefined;
		length?: number | undefined;
		size?: number | undefined;
	};

	export let raw = (
		strings: TemplateStringsArray,
		...placeholders: tg.Args<string>
	): tg.Blob.Builder => {
		return new tg.Blob.Builder(true, strings, ...placeholders);
	};
}
